### PR TITLE
Drop Ruby 2.3 and RubyGems 2.5

### PR DESCRIPTION
**Pull request by hsbt**
# Description:

This pr is migrated from https://github.com/rubygems/bundler/pull/7588

We will merge this after 2019-03-31.
______________

# Tasks:

- [x] Describe the problem / feature
- [x] Write tests
- [x] Write code to solve the problem
- [ ] Get code review from coworkers / friends

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

### CODE REVIEW

**Comment by deivid-rodriguez** [comment]
We should try to deal with #2575 and #3089 first, if possible.

**Comment by deivid-rodriguez** [comment]
As I said, I believe we should try to address #2575 and #3089 first, so I'm :-1: on doing this right now. Something we could do as a middle ground to reduce the number of jobs is to only test against the default rubygems version when testing bundler against ruby 2.3 and ruby 2.4.

**Comment by hsbt** [comment]
I understood the current behavior of `required_ruby_version` is unfriendly for users. 

But we couldn't backport the feature of `required_ruby_version` handling for Ruby 2.3 and its rubygems version. So, Ruby 2.3 users couldn't use it in the feature if we fixed its issues.

Do you have an any idea about this?

**Comment by deivid-rodriguez** [comment]
Yeah, you are right. Even if we fix this, the fix won't be backported to the rubies that currently need it. So the fix will only be effective for users who are currently doing `gem update --system` in CI. Also, I expect [ruby/setup-ruby](https://github.com/ruby/setup-ruby] to have alleviated this problem.

Still, I feel this is more of a philosophical issue. When we drop support and this issue affects people, I'd like to be able to tell people: "we have fixed this in rubygems. Make sure you update rubygems before running `gem install bundler` and the issue we'll be gone" rather than "you need to drop support for ruby 2.3" or "you need to stay on an old bundler version".

Because the first alternative encourges adoption of newer versions of bundler & rubygems while the latter discourages it.

That said, I understand that if we don't manage to fix the `required_ruby_version` issue soon, we'll need to drop support at some point. My take to dropping support for old rubies requires two conditions:

* The version being dropped has reached its end of life.
* Supporting the version is getting in the middle and hurting development.

So the question is whether the answer to the second question is already "Yes". I haven't personally been bothered by supporting ruby 2.3 and ruby 2.4 (yet). The thing that might start bothering me is the excessive number of CI jobs, but the bottleneck for CI is currently Windows + ruby 2.6, and we could stop testing for _some_ rubygems+bundler combinations without needing to drop full support.

So, to sum up my opinion: I'd like to wait a few more months and reevaluate this again by then.

**Comment by hsbt** [comment]
I see. This changes is not rush to merge. I hope to merge in this year at least.

**Comment by deivid-rodriguez** [comment]
Alright, so I was considering whether it makes sense to drop support now, and I believe it's best to merge this PR next December right before releasing the next minor version. We could merge it now, but I think it would become counterproductive, because:

* It will complicate backports to the stable branch, that still needs to support old rubies.
* It might allow incompatible changes into master that will later have to be addressed in the stable branch at release time.

I will keep rebasing this PR once in a while, and we can move on with this right before cutting the next stable branch. How does that sound?

**Comment by deivid-rodriguez** [comment]
Also, I want to proposed, since we are at it, to also drop support for 2.5 and 2.6 at the same time since they will be already unsupported by the end of this year.

**Comment by hsbt** [comment]
Ruby 2.6 is EOL today. I submit another PR to remove supporting under the Ruby 2.6.
